functionary/train_vision/train.py

```diff
@@ -0,0 +1,539 @@
+    cache_dir: Optional[str] = field(default=None)
+    optim: str = field(default="adamw_8bit")
+    model_max_length: int = field(
+        default=4096,
+        metadata={
+            "help": "Maximum sequence length. Sequences will be right padded (and possibly truncated)."
+        },
+    )
+    keep_assistant_prefix: bool = field(
+        default=False,
+        metadata={
+            "help": "Whether to mask the assistant prefix `<|from|>assistant\n<|recipient|>` during training"
+        },
+    )
+    prompt_template_version: str = field(
+        default="v2", metadata={"help": "choose prompt template to use for training"}
+    )
+    log_train_metrics: bool = field(
+        default=False,
+        metadata={"help": "set this true to log training metrics during training"},
+    )
+
+
+def trainer_save_model_safe(trainer: transformers.Trainer):
+    """Saves the model in fsdp.FULL_STATE_DICT mode to have the model weights
+    in .bin file format which is loadable by HF Transformers"""
+    if trainer.accelerator.state.fsdp_plugin.state_dict_type.name != "FULL_STATE_DICT":
+        trainer.accelerator.state.fsdp_plugin.set_state_dict_type("FULL_STATE_DICT")
+    trainer.save_model()
+
+
+def initialize_tokenizer(
+    *,
+    model: transformers.AutoModelForCausalLM,
+    model_name_or_path: str,
+    prompt_template: PromptTemplate,
+    model_max_length: int,
+    cache_dir: str,
+):
+    """Initialize tokenizer and add special tokens, resizing vocab and embedding"""
+    # Mistral requires left padding due to the Sliding Window Attention mechanism
+    if "mistral" in type(model).__name__.lower():
+        print("model is mistral so padding_side=left")
+        padding_side = "left"
+    else:
+        padding_side = "right"
+
+    # note that must set legacy=True, read more: https://github.com/huggingface/transformers/issues/25176
+    tokenizer = AutoTokenizer.from_pretrained(
+        model_name_or_path,
+        cache_dir=cache_dir,
+        model_max_length=model_max_length,
+        padding_side=padding_side,
+        legacy=True,
+    )
+
+    # Add special tokens
+    tokenizer.pad_token = tokenizer.eos_token
+    added_tokens = prompt_template.get_additional_tokens()
+    special_tokens = {"additional_special_tokens": added_tokens}
+    num_new_tokens = tokenizer.add_special_tokens(special_tokens)
+
+    # add chat_template for tokenizer
+    tokenizer.chat_template = prompt_template.get_chat_template_jinja()
+    print("tokenizer: ", tokenizer)
+
+    # Resize embedding
+    model.resize_token_embeddings(len(tokenizer))
+    if num_new_tokens > 0:
+        input_embeddings = model.get_input_embeddings().weight.data
+        output_embeddings = model.get_output_embeddings().weight.data
+
+        input_embeddings_avg = input_embeddings[:-num_new_tokens].mean(
+            dim=0, keepdim=True
+        )
+        output_embeddings_avg = output_embeddings[:-num_new_tokens].mean(
+            dim=0, keepdim=True
+        )
```

Comment:
Why resize is needed for this model? Could you add an explanation

Reply:
Oh this is **just in case we add new tokens to the tokenizer** In the future we may add new tokens, currently we don't. If we don't add new tokens, this chunk of code has no effect